### ISSUE TITLE
Restore checkpoint with custom environment & model 

### ISSUE THREAD

**Issue by esham01** [ray-project/ray#34155]
### What happened + What you expected to happen

Hi! I have checkpoints from a previous run of PPO with my custom environment and model using r-lib. I'm trying to load one of these checkpoints to further train them. I follow the algo.restore(path-to-checkpt) script I pasted below but for some reason am not able to load the checkpoint. I receive no bugs but when it runs, the loss I get is not the same or close to the checkpoint I am loading from.  The loss is the same as the loss I get when I start training from scratch. 

### Versions / Dependencies

ray                       2.3.1 

### Reproduction script

#Code for loading checkpoint
```
ModelCatalog.register_custom_model("my_torch_model", CustomTorchModel)
ray.init(num_gpus=1)
config = (
        get_trainable_cls("PPO")
        .get_default_config()
        # or "corridor" if registered above
        .environment(CustomEnv, env_config={})
        .framework("torch")
        .rollouts(num_rollout_workers=0)
        .training(
            model={
                "custom_model": "my_torch_model",
                # Extra kwargs to be passed to your model's c'tor.
                "custom_model_config": {},
            }
        )
        .resources(num_gpus=1)
    )

```

```
my_new_ppo = config.build()


my_new_ppo.restore(path_to_checkpoint)


my_new_ppo.train()`
```

# Code for saving checkpoint: 

```
algo = ppo.PPO(env="combat_spider_ppo", config={
    "framework": "torch",
    "disable_env_checking": True,
    "num_gpus": 1,
    "num_workers":0,
    "simple_optimizer":True,
    "model": {
        "custom_model": "my_torch_model",
        # Extra kwargs to be passed to your model's c'tor.
        "custom_model_config": {},
    },
})
```

```
for i in range(100):
    result = algo.train()
    print(pretty_print(result), flush=True)

    if i % 5 == 0:
        checkpoint_dir = algo.save()
        print(f"Checkpoint saved in directory {checkpoint_dir}")
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by Rohan138** [comment]
Could you provide more details about your custom model and/or env? Without more information, it's hard to reproduce/debug losses during training.

**Comment by esham01** [comment]
Thanks! For my custom env, it returns a float for the reward and this is the observation space that I have: 

```
dict = {}
dict["rgb"] = gymnasium.spaces.Box(low=-80, high=80, shape = (512,))
dict["compass"] = gymnasium.spaces.Box(low = -60, high =60, shape=(4,))
dict["gps"] = gymnasium.spaces.Box(low = -80, high = 80, shape=(3,))


dict["voxels"] = gymnasium.spaces.Box(low=0, high=26, shape=(27,))

dict["prompt"] = gymnasium.spaces.Box(low=-50, high= 50, shape = [512,])

self.observation_space = gymnasium.spaces.Dict(dict)
```

I also pasted some code from my custom model below: 

```
class CustomTorchModel(TorchModelV2, nn.Module):
    def __init__(self, obs_space, action_space, num_outputs, model_config, name):
        TorchModelV2.__init__(
            self, obs_space, action_space, num_outputs, model_config, name
        )
        nn.Module.__init__(self)

        self._obs_= []


        cfg = OmegaConf.load("conf.yaml")

        feature_net_kwargs = cfg.feature_net_kwargs

        feature_net = {}
        for k, v in feature_net_kwargs.items():
            v = dict(v)
            cls = v.pop("cls")
            cls = getattr(F, cls)
            feature_net[k] = cls(**v, device=device)

        feature_fusion_kwargs = cfg.feature_fusion

        feature_net = SimpleFeatureFusion(
        feature_net, **feature_fusion_kwargs, device=device)
        
        self._actor_ = actor = MultiCategoricalActor(
        feature_net,
        action_dim=[3, 3, 4, 25, 25, 8],
        device=device,
        **cfg.actor)

        self._mine_agent_ = MineAgent(actor=actor).to(device)
        self._value_ = None
        
    @override(TorchModelV2)
    def forward(self, input_dict, state, seq_lens):


        obs = input_dict["obs"]
        self._obs_ = obs

       
        dic = {
        "compass": obs["compass"],
        "gps": obs["gps"],
        "voxels": obs["voxels"].cpu().int(),
        "prompt": obs["prompt"],
        "rgb": obs["rgb"],
        }
        
        batch_obs = Batch(obs=dic)
        batch_val, self._value_ = self._mine_agent_(batch_obs)

        logits = batch_val.logits


        # masking out invalid actions 
        logits[:,-6] = -1e6
        
        logits[:,-4:] = -1e6

        x,_ = logits.shape

        zeros = torch.zeros((x, 280))

        logits_final = torch.cat((logits, zeros), dim=1)
    
        return logits_final, []

    @override(TorchModelV2)
    def value_function(self):
        return self._value_
    




class MineAgent(nn.Module):
    def __init__(
        self,
        actor: nn.Module,
        deterministic_eval: bool = True,
    ):
        super().__init__()
        self.actor = actor
        self._deterministic_eval = deterministic_eval
        self.dist_fn = actor.dist_fn

    def forward(
        self,
        batch: Batch,
        state=None,
        **kwargs,
    ) -> Batch:
        logits, hidden,value = self.actor(batch.obs, state=state)

        
        if isinstance(logits, tuple):
            dist = self.dist_fn(*logits)
        else:
            dist = self.dist_fn(logits)
        if self._deterministic_eval and not self.training:
            act = dist.mode()
        else:
            act = dist.sample()
        return Batch(logits=logits, act=act, state=hidden, dist=dist), value
    


class MultiCategoricalActor(nn.Module):
    def __init__(
        self,
        preprocess_net: nn.Module,
        *,
        action_dim: list[int],
        hidden_dim: int,
        hidden_depth: int,
        activation: str = "relu",
        device,
    ):
        super().__init__()
        self.mlps = nn.ModuleList()
        self.preprocess = preprocess_net
        for action in action_dim:
            net = build_mlp(
                input_dim=preprocess_net.output_dim,
                output_dim=action,
                hidden_dim=hidden_dim,
                hidden_depth=hidden_depth,
                activation=activation,
                norm_type=None,
            )
            self.mlps.append(net)
        self._action_dim = action_dim
        self._device = device

        self._lin_layer = nn.Sequential(nn.ReLU(), nn.Linear(68,68), nn.ReLU(), nn.Linear(68,1))

    def forward(self, x, state=None, info=None):
        hidden = None

        x, _ = self.preprocess(x)


        value = torch.cat([mlp(x) for mlp in self.mlps], dim=1)

        value = self._lin_layer(value).squeeze(1)
        
        return torch.cat([mlp(x) for mlp in self.mlps], dim=1), hidden, value

    @property
    def dist_fn(self):
        return lambda x: MultiCategorical(logits=x, action_dims=self._action_dim)
```

**Comment by Rohan138** [comment]
Could you compile your code into a single repro script that we can run to debug?
